docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,220 @@
+    [size=len] char* buf;
+};
+```
+
+The `len` variable specified to the `size` attribute represents the size (in bytes) of the `buf`.
+Refer to the [Grammar documentation](https://github.com/openenclave/oeedger8r-cpp/blob/master/docs/Grammar.md) for
+more information regarding the syntax of oeedger8r.
+
+To pass the `Blob` struct in an ECALL/OCALL, the oeedger8r currently supports the following two cases:
+- `in` parameter
+  ```
+  // EDL definition
+  void foo([in, count=1] Blob *b);
+  ```
+
+  The above code snippet shows an example of passing `b` as an `in` parameter. In this case,
+  the caller is responsible for initializing `b`, including `b->len` and the memory pointed by `b->buf`.
+  The callee can then read the content of `b`. However, any callee's modifications to
+  `b` will not reflect on the caller site (i.e., `b` is effectively read-only to the callee).
+
+- `inout` parameter
+  ```
+  // EDL definition
+  void foo[in, out, count=1] Blob *b);
+  ```
+
+  The above definition is an example of passing `b` as an `inout` parameter. Similar to the case
+  of the `in` parameter, the caller is responsible for initializing `b` and the callee can read
+  the content of the `b`. Besides, callee's modifications to `b` will now reflect on the caller
+  site with a limitation that the callee cannot re-allocate or resize `b->buf`
+  (which could lead to undefined behaviors).
+
+
+The case that remains unsupported is the `out` parameter as shown in the following example.
+```
+// EDL definition
+void foo[out, count=1] Blob *b);
+```
+The major difference between the `out` parameter and the other two is that caller is no longer responsible
+for initializing `b` (i.e., the caller can simply pass a `b` pointer with `b->buf` equals to `NULL`).
+Instead, the callee takes charge of initializing `b`, which implies that `b->buf` is variable-length.
+Such support is required by typical scenarios in which the caller cannot determine the size of `b->buf`.
+
+Proposed Design
+-----------
+
+An important constraint that affects the design of the `out` parameter support is the availability of
+shared memory between a host and an enclave. Given that this support may not be general across TEEs,
+we discuss two designs in this section.
+
+The following code snippet serves as a common example that both two design aims to support.
+- Host
+  ```
+  Blob b;
+  // b->len and b->buf have not been initialized yet.
+  foo(enclave, &b);
+  // b->len and b->buf have been initialized.
+  // b->len == 10 and b->data equals to data.
+  ```
+
+- Enclave
+  ```
+  // User-defined code
+  void foo(Blob *b)
+  {
+    b->len = 10;
+    b->buf = malloc(b->size);
+    memcpy(b->buf, data, b->size);
+  }
+  ```
+
+### Design with Shared Memory
+
+With the support of shared memory, the design is more straightforward.
+The core idea of the design is that after the callee initializes the `b`,
+it also allocates a host memory and copies the content `b` to the allocated
```

Comment:
Line 29 talks about both directions (ECALL and OCALL) but here only talks about one direction (ECALL).
What about out parameters of OCALLs?

Reply:
Updated.